workers/unity/Packages/com.improbable.gdk.core/GameObjectRepresentation/GameObjectDispatcherSystem.cs

```diff
@@ -0,0 +1,161 @@
+                gameObjectComponentDispatcher.AuthoritiesChangedComponentGroup =
+                    GetComponentGroup(gameObjectComponentDispatcher.AuthoritiesChangedComponentTypes);
+                if (gameObjectComponentDispatcher.ComponentsUpdatedComponentTypes.Length > 0)
+                {
+                    gameObjectComponentDispatcher.ComponentsUpdatedComponentGroup =
+                        GetComponentGroup(gameObjectComponentDispatcher.ComponentsUpdatedComponentTypes);
+                }
+
+                if (gameObjectComponentDispatcher.EventsReceivedComponentTypeArrays.Length > 0)
+                {
+                    gameObjectComponentDispatcher.EventsReceivedComponentGroups =
+                        new ComponentGroup[gameObjectComponentDispatcher.EventsReceivedComponentTypeArrays.Length];
+                    for (var i = 0; i < gameObjectComponentDispatcher.EventsReceivedComponentTypeArrays.Length; i++)
+                    {
+                        gameObjectComponentDispatcher.EventsReceivedComponentGroups[i] =
+                            GetComponentGroup(gameObjectComponentDispatcher.EventsReceivedComponentTypeArrays[i]);
+                    }
+                }
+
+                if (gameObjectComponentDispatcher.CommandRequestsComponentTypeArrays.Length > 0)
+                {
+                    gameObjectComponentDispatcher.CommandRequestsComponentGroups =
+                        new ComponentGroup[gameObjectComponentDispatcher.CommandRequestsComponentTypeArrays.Length];
+                    for (var i = 0; i < gameObjectComponentDispatcher.CommandRequestsComponentTypeArrays.Length; i++)
+                    {
+                        gameObjectComponentDispatcher.CommandRequestsComponentGroups[i] =
+                            GetComponentGroup(gameObjectComponentDispatcher.CommandRequestsComponentTypeArrays[i]);
+                    }
+                }
+            }
+        }
+
+        protected override void OnUpdate()
+        {
+            foreach (var gameObjectComponentDispatcher in GameObjectComponentDispatchers)
+            {
+                gameObjectComponentDispatcher.InvokeOnAddComponentLifecycleCallbacks(this);
+                gameObjectComponentDispatcher.InvokeOnRemoveComponentLifecycleCallbacks(this);
+                gameObjectComponentDispatcher.InvokeOnAuthorityChangeLifecycleCallbacks(this);
+            }
+
+            foreach (var spatialOSBehaviourManager in spatialOSBehaviourManagers)
+            {
+                    spatialOSBehaviourManager.EnableSpatialOSBehaviours();
+            }
+
+            foreach (var gameObjectComponentDispatcher in GameObjectComponentDispatchers)
+            {
+                gameObjectComponentDispatcher.InvokeOnAuthorityChangeUserCallbacks(this);
+                gameObjectComponentDispatcher.InvokeOnComponentUpdateUserCallbacks(this);
```

Comment:
We might even need to keep and ordering list on each entity. A list of enums saying what arrived in what order.

Reply:
Or we could squash. Or we could assume auth loss comes before updates and auth gained comes after

Reply:
Under which circumstances would you want e.g. component update callbacks to fire before authority callbacks? Agree though that the overall approach here looks fishy e.g. for situations where user code depends on an event being processed before component field updates.

Reply:
Ok, after thinking about this a little more, here is what I'm currently think:
- If we want to preserve the order of the incoming ops, that would require a refactor of the GDK Core as the GDK Core currently does not surface that information. A such, those changes shall not be in scope of this PR.
- Assuming that the information about the order of incoming ops can not be recovered given the current limitations of GDK Core, what we have here is as good as any other possible alternative.

Reply:
We definitely can surface that info - but I agree it is out of scope for this PR

Reply:
You want the updates before the authority callbacks if you gained authority. Doing it the other way round is wrong.

Reply:
This stuff doesn't need any fancy things in the core. I mentioned as an extension you could want to preserve orderings in more complete ways but even without that you need to order the updates and auth callbacks correctly. As it currently stands this code is wrong.

Reply:
Re: "You want the updates before the authority callbacks if you gained authority": Why? Wouldn't you want to maybe do some sort of setup upon gaining authority that subsequent component updates depend on?
Re: "I mentioned as an extension you could want to preserve orderings in more complete ways but even without that you need to order the updates and auth callbacks correctly": How would I do that?

Reply:
For the first question: No you don't want to be receiving updates from other workers during a block where you expect to have authority. The way a user expects to interpret update callbacks on writers is as something that worker sent itself. You probably want to get the current state of the component, and you might want to access the updates that have been received before the auth change (for predictive code etc) but you don't want to be told you received an update after the auth change. That would cause users to have to write all their writer callback code having to check if they just gained authority or had it for a while :S.
The ECS layer will give you the ability to access the updates that were received before hand if you want them. You could also do it by having a companion behaviour, which only has a reader, so it gets activated earlier and will receive the component update before the auth callback. But that is a special case where a user wants to do something clever, so it can be a slightly more awkward flow. The standard case is that once you have authority you don't care what happened before.

For the second point: That was the extension that was meant for another time. E.g. an list of enums per component. So I agree with your assessment that this is out of scope for now. More something to think about for the future.

Reply:
Re: first point: Not sure if I understand your point but if we don't want to receive updates on authoritative workers then that's solved in the GDK Core, as in the GDK Core should or shouldn't emit the reactive component tags. Assuming that reactive component tags are not emitted (afaik they currently are which I'm fine with), the execution order won't matter as there are no tags to look for.

Reply:
No, you still want to be able to find what the updates were for some cases. You just don't want them as callbacks.So in this case we can't assume the gdk core is removing them.